### PR TITLE
Add GUI tests for ML/AI plugin

### PR DESCRIPTION
**Pull request by lbianchi-lbl**
## Summary/Motivation

- Add tests and CI configuration so that ML/AI plugin can be tested as part of the CI checks

## Changes proposed in this PR:

- Add test module `test_ml_ai.py` running GUI tests for ML/AI plugin using the flowsheet loaded from the ML/AI example FOQUS session file
- Implement changes to pytest fixtures and `pytest_qt_extras` required so that the FOQUS session and main window is reinitialized upon entering each test module (as opposed to 
  - This is being added here as `test_ml_ai.py` is the first test module being added that requires a different session being loaded during a test run (previously, a single session was enough to run all the GUI tests in `test_uq.py`)
- Set up optional dependency `tensorflow` in the CI workflow so that it is installed in a subset of the checks, and skipped in the tests if not importable
  - This aims to test both scenarios where users installed `tensorflow` or not

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the copyright and license terms described in the LICENSE.md file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: .github/workflows/checks.yml [hunk 1]
```diff
@@ -72,6 +73,8 @@ jobs:
             os-version: windows-2019
           - python-version: "3.8"  # avoid uploading coverage for full matrix
             use_coverage: true
+          - python-version: "3.8"  # this is to avoid installing optional dependencies in all environments
```
Comment (bpaul4) [inline .github/workflows/checks.yml:76]: Does this mean that these tests will not work or that users will not automatically install TensorFlow when installing FOQUS in Python versions other than 3.8?

File: foqus_lib/conftest.py [hunk 2]
```diff
@@ -48,31 +49,46 @@ def psuade_path():
...
+        session.makeWorkingDirStruct()
+        session.makeWorkingDirFiles()
 
-    dat = session.session(useCurrentWorkingDir=True)
-    dat.foqusSettings.psuade_path = str(psuade_path)
-    return dat
+        dat = session.session(useCurrentWorkingDir=True)
+        dat.foqusSettings.psuade_path = str(psuade_path)
+        yield dat
+    # TODO is there any cleanup to be done, considering that the directory will be changed?
```
Comment (bpaul4) [inline foqus_lib/conftest.py:94]: If FOQUS closes each session before opening a new one (generalizing to an arbitrary number of sessions), then the only "leftover" files would be generated pycache or log files. I don't believe those would be problem, especially if the test client cleans/remakes the directory on each test run.

File: foqus_lib/framework/uq/Common.py [hunk 3]
```diff
@@ -66,7 +66,6 @@ def showError(self, error, out=None):
                 )
                 if out is not None:
                     msgBox.setDetailedText(out)
-                msgBox.exec_()
```
Comment (bpaul4) [inline foqus_lib/framework/uq/Common.py:69]: Does removing this potentially hide error messages generated by non-Plugin tools? I'm wondering if we need to keep this

File: foqus_lib/gui/tests/test_ml_ai.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    def statusbar_message(self, main_window: mainWindow) -> str:
+        return main_window.statusBar().currentMessage()
+
+    def test_flowsheet_run_successful(
+        self,
+        trigger_flowsheet_run_action,
+        statusbar_message: str,
+        text_when_success: str = "Finished Single Simulation... Success",
+    ):
+        assert text_when_success in statusbar_message
```
Comment (bpaul4) [inline foqus_lib/gui/tests/test_ml_ai.py:149]: This all looks very nice, excellent job!

**Review by bpaul4 — COMMENTED** [review]
I have a couple of comments, but overall it looks very good and is well written.

File: .github/workflows/checks.yml (hunk 1 above)
Comment (lbianchi-lbl) [reply .github/workflows/checks.yml:76]: This won't affect users, and is only a directive to specify that, in the environment where CI jobs run, `tensorflow` should only be installed for a subset of the combinations. Currently, there are 3 Python versions x 3 OSes = 9 total combinations, but the GUI tests only work for 3.7 and 3.8, so specifying `tensorflow` only for Python 3.8 is a way to test both "with TF" and "without TF" using jobs with Python 3.8 and 3.7 respectively.

File: foqus_lib/conftest.py (hunk 2 above)
Comment (lbianchi-lbl) [reply foqus_lib/conftest.py:94]: I agree, my impression was also that creating a working directory would take care of most things. However, there is some additional cleanup being done in e.g. `mainWindow.closeEvent()` that the GUI tests currently need to bypass, so I've left a note to remind me/us to look into it down the line.

File: foqus_lib/framework/uq/Common.py (hunk 3 above)
Comment (lbianchi-lbl) [reply foqus_lib/framework/uq/Common.py:69]: This is a valid point. Unfortunately, using modal dialogs such as `QMessageBox` through the `exec_()` method causes all sorts of issues with automated GUI testing. More specifically this has to do with Qt's event loop: basically calling `exec_()` halts the execution of the main event loop in which the main window is running, and starts a separate event loop for the modal dialog. Once the modal dialog is dismissed, the main event loop resumes. This is fine during interactive use; however, the infrastructure used for automated testing (the "qtbot") is also running in the main event loop, and is therefore "trapped" whenever a modal dialog is created with `exec_()`. I had implemented a workaround that seemed to be working, however this started causing issues when adding the `test_ml_ai.py` file.

Long story short, getting rid of this particular modal is a temporary fix to get this set of tests to work on CI. I'm already working on a more robust solution that will encapsulate user notifications with a PyQt signal, that can be connected to display a modal dialog for interactive use, or inspected directly (with the modal dialogs turned off) for automated testing.

File: foqus_lib/gui/tests/test_ml_ai.py (hunk 4 above)
Comment (lbianchi-lbl) [reply foqus_lib/gui/tests/test_ml_ai.py:149]: Thanks! Once we have the notification signal set up, we'd probably want to use that for this assertion as it is slightly more robust (the status bar is a global object and might be modified by some other part of the GUI before we check, resulting in a false-positive test failure).

**Review by bpaul4 — APPROVED** [review]
Your comments make sense and the updates look great, so I will go ahead and merge this.
